### ISSUE TITLE
Investigate cloud based VNC access

### ISSUE THREAD

**Issue by FreddieAkeroyd** [ISISComputingGroup/IBEX#5507]
As an instrument scientists I would like to be able to offer read-only or read/write vnc access to users for the duration of their experiment. This access should not require VPN.  A cloud based VNC solution would likely be appropriate.
This will help implement  #5501 and #5502
 
## acceptance criteria

- [x] instrument scientists able to grant and and revoke vnc access to user (to limit to duration of their experiment)
- [x] instrument scientist able to choose read only or read/write access
- [x] adding users and pc's to relevant VNC groups understood to make user management easier
- [x] user authentication method approved with site security
- [x] Documentation for support and install

**Comment by John-Holt-Tessella** [comment]
@ChrisM-S has setup some machine, talk to himn to find out how to access them and what testing is left to do.

**Comment by John-Holt-Tessella** [comment]
Muon and LARMOR need to be included in the trial (emphisis on this being a trial and consider when to do this before or after security say ok)

**Comment by davidkeymer** [comment]
### Proposed structure for VNC Cloud Groups:

![VNC Cloud Groups](https://user-images.githubusercontent.com/10550207/86485047-9e0dec00-bd4f-11ea-8802-285f56966a20.png)

### Proposed procedure for setting up VNC Cloud access:

**Owner / Admin (Single Experiment Controls group member)**

1. Creates "Machine Groups" in VNC portal all within the ISIS "Team", one per physical ISIS instrument (See diagram above)
1. Invites Instrument Scientists (IS) to create a VNC account via VNC portal
1. Grants "Manager" privilege to IS
1. Adds IS to appropriate Machine (instrument) Group(s)

**Manager (IS or other ExptCtrl member)**

1. Creates VNC Cloud account using link in invitation email from Owner/Admin
1. Sends emails to external users via VNC portal inviting them to create a VNC account
1. Grants "User" privilege to users
1. Adds users to appropriate Machine (instrument) Group(s)
1. Adds "local" computers (viewing, analysis, etc.) to Machine Group(s) (See diagram above)
1. (Optional if willing & able) Installs and configures VNC server on selected computers
1. Performs below steps to use VNC client

**User (External facility user):**

1. Creates VNC cloud account using link in invitation email from IS
1. Downloads, installs and runs VNC client.
1. Logs in and is presented with list of machines authorised to connect to
1. Connects to a machine, typically a general access cabin PC (NDCxxx) or analysis machine (NDLxxx)
_[at this point has same access as if physically present in instrument cabin]_
1. Connects to instrument control computer (NDXxxx) via RDP (if session not already established)

**Comment by FreddieAkeroyd** [comment]
That looks really good. Could you just clarity how read only/read-write access is done. Is that specified at machine/group level, user vnc level, or via additional user domain login?

**Comment by davidkeymer** [comment]
Thanks for having a look, Freddie.  Good to know we're on the right track!  The method of granting read/write access is still to be finalised as there are several options.  I've asked @ChrisM-S for further thoughts which the three of us will probably need to discuss.

**Comment by davidkeymer** [comment]
It has been proposed that the _Read Only_ and _Read/Write_ access be controlled by changing the privileges on the standard viewing machine's user account (NDCxxx\Inst...).

This will be for the simplest case for when the viewing machine is running Windows and has (ISIS) standard local user and admin accounts, although the principal still applies to a Linux analysis machine and other cabin machines.

1. Instrument scientist changes permissions in VNC server settings for "standard user account" to be either R or R/W to suit experiment
1. User logs in to VNC Cloud using their personal account
1. User connects via VNC client to instrument viewing machine using "standard user account"
1. User able to view only or control experiment according to permissions set
1. Instrument scientist sets privileges back to Read Only after experiment ends

See [VNC article](https://help.realvnc.com/hc/en-us/articles/360002253618-Managing-Users-and-Session-Permissions-for-VNC-Server) for more information.

**Comment by davidkeymer** [comment]
It is believed that all acceptance criteria have now been met.  However, to date, it has not been possible to test the Linux client and server applications because of the unavailability of instrument analysis machines after the recent site power cut.  One of the ExptCtrlGrp's machines may be able to be used for this instead.

The reviewer will need a VNC Cloud account which e.g. CM-S can create, or at least send reviewer an invitation to create one.

Suggest the reviewer go though the procedures above on a Windows machine to check that the principle is sound.  An instrument "viewing" PC may be available, or one of the Group's test machines may suit.  Reviewer may wish to approve this ticket if all is well, and then create another to test on Linux systems (or reject with rework for this) or perhaps even test it themselves with a group machine.

#### NB

As this is an "investigation", the instructions are not "click by click" and only give an overview of what needs to be done in each step (The VNC Cloud admin portal is fairly self-explanatory and links to detailed articles are provided above).  It is envisaged that a follow-up ticket will be created to actually produce detailed instructions for the Instrument Scientists to follow, using these as a basis.

**Comment by JamesKingWork** [comment]
Where is the site security advice documented? Was it on an email or on the intranet or something?

**Comment by JamesKingWork** [comment]
The procedures look good, I think it would be useful to have directions (a combination of the procedures here and VNC articles which look quite useful) for instrument scientists on the user manual or other relevant area. Can we also have the procedures in the dev manual somewhere please.

**Comment by DominicOram** [comment]
Re-pointing as only got review left

**Comment by JamesKingWork** [comment]
Impeded waiting on a reply from RealVNC about using only UK data centres for site security

**Comment by davidkeymer** [comment]
> Where is the site security advice documented? Was it on an email or on the intranet or something?

I believe it was in a reply to a query that @ChrisM-S sent to site security (unsure of official group title).  Perhaps he could comment further...

**Comment by JamesKingWork** [comment]
> > Where is the site security advice documented? Was it on an email or on the intranet or something?
> 
> I believe it was in a reply to a query that @ChrisM-S sent to site security (unsure of official group title). Perhaps he could comment further...

Yes, this was discussed on teams

**Comment by kjwoodsISIS** [comment]
@JamesKingWork , @davidkeymer - Is the discussion on Teams stored somewhere permanent (e.g. the Sharepoint)?  I don't think Teams is an appropriate place to store the material, because it will soon scroll off the top and be forgotten about.  This ticket is not an appropriate place to store the information either - we don't want security information posted in GitHub.  I think Sharepoint or similar is a better place to store the information.  You can add a link to it.

**Comment by JamesKingWork** [comment]
@kjwoodsISIS Yes it is also on a document in sharepoint

**Comment by kjwoodsISIS** [comment]
@JamesKingWork  - my point was that the link to the Sharepoint document should be in this ticket.  If the link exists only in Teams, it is going to be pretty difficult to find it if we ever need to refer to it again.

**Comment by JamesKingWork** [comment]
I see what you mean: https://stfc365.sharepoint.com/sites/IBEX/_layouts/15/Doc.aspx?OR=teams&action=edit&sourcedoc={C6C84B27-3728-46B3-BE7E-AFD413853934}

**Comment by kjwoodsISIS** [comment]
@davidkeymer  - is this issue impeded any longer?  (I am guessing not, given that we are about to send out instructions for using cloud-based VNC).

**Comment by davidkeymer** [comment]
@Kevin-Woods-Tessella No, no reason to be impeded.  I'd misremembered that this ticket was impeded because of lack of testing on Linux, and had forgotten that we're assuming this will work when eventually tried.

I shall remove the label and put the ticket in for review...

**Comment by davidkeymer** [comment]
Instructions for both instrument scientists and external users placed in this [Sharepoint area](https://stfc365.sharepoint.com/:f:/r/sites/IBEX/Shared%20Documents/General/Remote%20Users%20Sep%202020?csf=1&web=1&e=qy9qLg)

**Comment by DominicOram** [comment]
Given the documentation linked from here I'm happy to call this complete.
